File: Week 02/id_063/C++/LeetCode_51_063.cpp
#include <vector>
#include <string>
#include <stdlib.h>
#include <iostream>
using namespace std;

class Solution {
public:

    void choice2slist(vector<int>& choice, vector<string>& v) {
        char* arr = new char[choice.size()];
        for (int i = 0; i < choice.size(); i++) {
            memset(arr, '.', choice.size() * sizeof(char));
            arr[choice[i]] = 'Q';
            v[i] = string(arr, choice.size());
        }


        delete[] arr;
    }

    void dfs(vector<vector<string>>& result, vector<int>& choice, vector<string>& print_info, int cur_level, int max_level) {
        if (cur_level == max_level) {
            choice2slist(choice, print_info);
            result.push_back(print_info);
        }

        int cur_j;
        for (cur_j = 0; cur_j < max_level; cur_j++) {

            bool conflict = false;
            for (int i = 0; i < cur_level; i++) {
                if (choice[i] == cur_j) {
                    conflict = true;
                    break;
                }

                if (abs(cur_level-i) == abs(cur_j - choice[i])) {
                    conflict = true;
                    break;
                }
            }

            if (conflict) {
                continue;
            }

            choice[cur_level] = cur_j;
            dfs(result, choice, print_info, cur_level + 1, max_level);
        }

    }

    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> v;
        vector<int> choice(n);
        vector<string> print_info(n);
        dfs(v, choice, print_info, 0, n);

        return v;
    }
};

File: Week 06/id_063/Leetcode_51_063.java
/*
思路

递归求解，没什么可说的，注意横竖和斜线上是否有皇后，进行减枝回溯即可
 */

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class Solution {

    private List<String> choice2str_list(int[] choice) {
        List<String> retList = new ArrayList<>(choice.length);
        String initString = "";
        for (int i = 0; i < choice.length; i++) initString = initString + ".";

        for (int queuePos : choice) {
            StringBuilder s = new StringBuilder(initString);
            s.setCharAt(queuePos, 'Q');
            retList.add(s.toString());
        }

        return retList;
    }

    private List<List<String>> dfs(List<List<String>> result, int[] choice, int curLevel, int maxLevel) {
        if (curLevel == maxLevel) {
            result.add(choice2str_list(choice));
            return result;
        }

        for (int cur_j = 0; cur_j < maxLevel; cur_j++) {
            boolean conflict = false;
            for (int i = 0; i < curLevel; i++) {
                // 列冲突
                if (choice[i] == cur_j) {
                    conflict = true;
                    break;
                }

                // 对角线冲突
                if (Math.abs(curLevel -i) == Math.abs(cur_j-choice[i])) {
                    conflict = true;
                    break;
                }
            }

            if (conflict) {
                continue;
            }

            choice[curLevel] = cur_j;
            dfs(result, choice, curLevel + 1, maxLevel);
        }

        return result;
    }

    public List<List<String>> solveNQueens(int n) {
        return dfs(new LinkedList<List<String>>(), new int[n], 0, n);
    }
}
